File: ruby/card_deck.rb
# frozen_string_literal: true
class CardDeck
  def initialize(cards = build_cards)
    @cards = cards
  end

  def each(&block)
    @cards.each(&block)
  end

  def size
    @cards.size
  end

  def shuffle!
    @cards.shuffle!
    self
  end

  def draw!
    @cards.shift
  end

  def cut!(num_of_cards)
    [
      CardDeck.new(@cards[0, num_of_cards]),
      CardDeck.new(@cards[num_of_cards, @cards.size])
    ]
  end

  def +(other_deck)
    cards_from_other = []
    other_deck.each do |card|
      cards_from_other << card
    end

    CardDeck.new(@cards + cards_from_other)
  end

  def ==(other_deck)
    cards_from_other_deck = []
    other_deck.each do |card|
      cards_from_other_deck << card
    end

    @cards == cards_from_other_deck
  end

  private

  def build_cards
    [].tap do |cards|
      %w(♠︎ ♣︎ ♥︎ ♦︎).each do |suite|
        %w(A 2 3 4 5 6 7 8 9 10 J Q K).each do |rank|
          cards << Card.new(suite, rank)
        end
      end
      2.times { cards << Card.new("*", "Joker") }
    end
  end
end

class Card
  attr_reader :suite, :rank

  def initialize(suite, rank)
    @suite = suite
    @rank = rank
  end

  def ==(other)
    suite == other.suite && rank == other.rank
  end
end

File: kotlin/src/main/kotlin/CardDeck.kt
class CardDeck private constructor(private val cards: List<Card>) : Iterable<Card> {
    companion object {
        fun build(): CardDeck {
            val cards = mutableListOf<Card>()
            val suites = setOf(Suite.CLUBS, Suite.DIAMONDS, Suite.HEARTS, Suite.SPADES)
            val ranks = setOf(
                Rank.ACE, Rank.TWO, Rank.THREE, Rank.FOUR, Rank.FIVE, Rank.SIX, Rank.SEVEN,
                Rank.EIGHT, Rank.NINE, Rank.TEN, Rank.JACK, Rank.QUEEN, Rank.KING
            )

            for (suite in suites) {
                for (rank in ranks) {
                    cards.add(Card(suite, rank))
                }
            }

            cards.add(Card(Suite.WILD, Rank.JOKER))
            cards.add(Card(Suite.WILD, Rank.JOKER))

            return CardDeck(cards)
        }
    }

    override fun equals(other: Any?): Boolean = when {
        other is CardDeck -> cards == other.cards
        else -> super.equals(other)
    }

    override fun iterator() = cards.iterator()
    override fun hashCode() = cards.hashCode()

    fun size() = cards.size

    fun shuffle(): CardDeck {
        return CardDeck(
            cards.toMutableList()
                .apply { shuffle() }
        )
    }

    fun draw(): Pair<Card, CardDeck> {
        val remainingCards = cards.toMutableList()
        val drawnCard = remainingCards.removeAt(0)
        return drawnCard to CardDeck(remainingCards)
    }

    fun cut(numberOfCards: Int) = Pair(
        CardDeck(cards.take(numberOfCards)),
        CardDeck(cards.drop(numberOfCards))
    )

    operator fun plus(otherDeck: CardDeck): CardDeck {
        return CardDeck(
            cards.toMutableList()
                .apply { addAll(otherDeck) }
        )
    }
}
